_posts/2020-09-16-hibernate-search.md

```diff
@@ -0,0 +1,440 @@
+              .matching(word)
+              .createQuery();
+
+    FullTextQuery fullTextQuery = fullTextEntityManager
+                          .createFullTextQuery(foodQuery, Post.class);
+    return (List<Post>) fullTextQuery.getResultList();
+  }
+  //.........
+  
+}
+```
+
+Let me explain the above code line by line:
+1. First, we create an object of `FullTextEntitiyManager` which is a wrapper over our `EntityManager`
+2. Next, we create `QueryBuilder` for the index on which we want to perform a search. We also need to pass the entity class object in it.
+3. Using `QueryBuilder` we finally build our `Query`. 
+4. Next, we make use of the keyword query `keyword()` which allows us to look for a specific word in a field or fields. Finally, we pass the word that we want to search in the `matching` 
+function.
+4. Lastly, we wrap everything in `FullTextQuery` and fetch result list by calling `getResultList()` function.
+
+One thing to note here is that although we are performing a query on Elasticsearch, Hibernate will still fire a query on the database to fetch the full entity. Which make sense, because as we saw in the previous section we didn't store all the fields of 
+`Post` entity and those fields still needed to be retrieved. If you only wanted to fetch what's stored in your index anyway and think this database call is redundant, I know just the way. Keep reading on to find out.
+
+Let's move to the next query. 
+
+Let's retrieve all the posts whose `likeCount` is greater than 1000 and should optionally contain 'food' hashtag and 
+'Literature' tag.
+
+```java
+public List<Post> getBasedOnLikeCountTags(Long likeCount, 
+                                          String hashTags, 
+                                          String tag){
+
+  FullTextEntityManager fullTextEntityManager = 
+                        Search.getFullTextEntityManager(entityManager);
+  QueryBuilder qb = fullTextEntityManager
+            .getSearchFactory()
+            .buildQueryBuilder()
+            .forEntity(Post.class)
+            .get();
+
+  Query likeCountGreater = qb.range()
+            .onField("likeCount")
+            .above(likeCount)
+            .createQuery();
+
+  Query hashTagsQuery = qb.keyword()
+            .onField("hashTags")
+            .matching(hashTags)
+            .createQuery();
+
+  Query tagQuery = qb.keyword()
+            .onField("tag")
+            .matching(tag)
+            .createQuery();
+
+  Query finalQuery = qb.bool()
+            .must(likeCountGreater)
+            .should(tagQuery)
+            .should(hashTagsQuery)
+            .createQuery();
+
+  FullTextQuery fullTextQuery = fullTextEntityManager
+                .createFullTextQuery(finalQuery, Post.class);
+  fullTextQuery.setSort(qb.sort().byScore().createSort());
+  return (List<Post>) fullTextQuery.getResultList();
+}
+```
+
+As you can see in the snippet above for `likeCount` we are using range query. Using only `above()` is equivalent to `>=` 
```

Comment:
"As you can see..." may turn some readers off if they don't see it at the first glance (which almost no one does, I bet). I changed the phrase.